app/controllers/api/messaging.php

```diff
@@ -0,0 +1,2138 @@
+<?php
+
+use Appwrite\Event\Delete;
+use Appwrite\Event\Messaging;
+use Appwrite\Extend\Exception;
+use Appwrite\Permission;
+use Appwrite\Role;
+use Appwrite\Utopia\Database\Validator\CustomId;
+use Appwrite\Utopia\Database\Validator\Queries\Messages;
+use Appwrite\Utopia\Database\Validator\Queries\Providers;
+use Appwrite\Utopia\Database\Validator\Queries\Subscribers;
+use Appwrite\Utopia\Database\Validator\Queries\Topics;
+use Appwrite\Utopia\Response;
+use Utopia\App;
+use Utopia\Database\Database;
+use Utopia\Database\Document;
+use Utopia\Database\Exception\Duplicate as DuplicateException;
+use Utopia\Database\Helpers\ID;
+use Utopia\Database\Query;
+use Utopia\Database\Validator\Authorization;
+use Utopia\Database\Validator\Datetime as DatetimeValidator;
+use Utopia\Database\Validator\UID;
+use Utopia\Validator\ArrayList;
+use Utopia\Validator\Boolean;
+use Utopia\Validator\JSON;
+use Utopia\Validator\Text;
+use Utopia\Validator\WhiteList;
+
+App::post('/v1/messaging/providers/mailgun')
+    ->desc('Create Mailgun Provider')
+    ->groups(['api', 'messaging'])
+    ->label('audits.event', 'providers.create')
+    ->label('audits.resource', 'providers/{response.$id}')
+    ->label('scope', 'providers.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_ADMIN, APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'messaging')
+    ->label('sdk.method', 'createMailgunProvider')
+    ->label('sdk.description', '/docs/references/messaging/create-mailgun-provider.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_CREATED)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_PROVIDER)
+    ->param('providerId', '', new CustomId(), 'Provider ID. Choose a custom ID or generate a random ID with `ID.unique()`. Valid chars are a-z, A-Z, 0-9, period, hyphen, and underscore. Can\'t start with a special char. Max length is 36 chars.')
+    ->param('name', '', new Text(128), 'Provider name.')
+    ->param('default', false, new Boolean(), 'Set as default provider.', true)
+    ->param('enabled', true, new Boolean(), 'Set as enabled.', true)
+    ->param('isEuRegion', false, new Boolean(), 'Set as EU region.', true)
+    ->param('from', '', new Text(256), 'Sender Email Address.')
+    ->param('apiKey', '', new Text(0), 'Mailgun API Key.')
+    ->param('domain', '', new Text(0), 'Mailgun Domain.')
+    ->inject('dbForProject')
+    ->inject('response')
+    ->action(function (string $providerId, string $name, bool $default, bool $enabled, bool $isEuRegion, string $from, string $apiKey, string $domain, Database $dbForProject, Response $response) {
+        $providerId = $providerId == 'unique()' ? ID::unique() : $providerId;
+
+        $provider = new Document([
+            '$id' => $providerId,
+            'name' => $name,
+            'provider' => 'mailgun',
+            'type' => 'email',
+            'default' => $default,
+            'enabled' => $enabled,
```

Comment:
should enabled be false until everything is populated?

Reply:
I don't think it will make any difference here, nothing will be persisted until the `createDocument` call anyway

Reply:
But there's nothing between here and the `createDocument()` call to validate or force `$enabled` to be false 🧐

Reply:
Not sure I understand, why would we need to do that?

Reply:
To prevent people from enabling a provider that doesn't have everything filled out yet.

Reply:
All other parameters are required

Reply:
According to the design, a developer needs to be able to start configuring a provider, save their progress, and go back to it to finish it at a later time.